lib/plugins/cors.js

```diff
@@ -73,66 +120,234 @@ function matchOrigin(req, origins) {
+        // is missing, or the resource requested for is missing, returna 400.
+        var reqOrigin = req.headers.origin;
+        var reqMethod = req.headers[AC_REQ_METHOD];
+        var reqHeaders = req.headers[AC_REQ_HEADERS];
+        var errMsg = 'invalid preflight request';
+        var errMsg2 = 'unsupported method';
 
-            if (opts.credentials) {
-                res.setHeader(AC_ALLOW_ORIGIN, origin);
-                res.setHeader(AC_ALLOW_CREDS, 'true');
-            } else {
-                res.setHeader(AC_ALLOW_ORIGIN, origin);
+        // 2) check that the incoming preflight headers are ok
+        if (!reqOrigin || !reqMethod || !reqHeaders) {
+            return next(new errs.BadRequestError(errMsg));
+        }
+
+        // 3.1) check if origin is valid and matched
+        var matchedOrigin = matchOrigin(reqOrigin, origins);
+
+        // 3.2) make sure headers are all allowed
+        var splitHeaders = reqHeaders.split(/\s*,\s*/);
+        var validHeaders = splitHeaders.every(function headerOk(h) {
+            return (allowHeaders.indexOf(h) !== -1);
+        });
+
+        // if either 3.1 or 3.2 failed, return 400.
+        if (!matchedOrigin || !validHeaders) {
+            return next(new errs.BadRequestError(errMsg));
+        }
+
+        // 3.4) check that the requested resource and method are valid. couple
+        // ways we can do this. user passes in an option that can either be:
+        //
+        //  - a restify server instance:
+        //      in this scenario, we use the router to 'simulate' a routing
+        //      request, which will populate res.methods with the supported
+        //      methods for that resource. this is an easy way to
+        //      automatically get full support without having to implement
+        //      server.opts on every single cors resource.
+        //
+        //      on a successful 'route', the router will populate
+        //      res.methods with the list of supported methods, and also
+        //      likely return a 405, but that's expected. iif router returns
+        //      a 404, that means the resource doesn't even exist, so we
+        //      need to send a 400 back.
+        //
+        //  - a function, that given req/res, returns an array of methods
+        //      a user provided function that will return an array of supported
+        //      methods given a req and res object.
+        //
+        //  - a value of false
+        //      don't do anything, user will manually implement server.opts for
+        //      every single resource.
+
+        var findMethodsFn;
+
+        if (typeof preflightStrategy === 'function') {
+            findMethodsFn = preflightStrategy;
+        } else if (typeof preflightStrategy === 'object' &&
+                   preflightStrategy.router &&
+                   preflightStrategy.router.find) {
+
+            var server = preflightStrategy;
+            findMethodsFn = function findMethods() {
```

Comment:
Should this accept `req, res, cb` args and invoke `cb` with `null, res.methods`?